File: platform/sdk/go/native_program_call.go
package layerx

import "encoding/binary"

type NativeProgramCall struct {
	ProgramID         [32]byte
	GuestABI          uint16
	Entrypoint        string
	Calldata          []byte
	Capabilities      []byte
	AccessDeclaration []byte
	ResponseCapacity  uint32
	Resources         [7]uint64
}

func EncodeNativeProgramCall(call NativeProgramCall) ([]byte, error) {
	if call.ProgramID == ([32]byte{}) || (call.GuestABI != 1 && call.GuestABI != 2) || len(call.Entrypoint) == 0 || len(call.Entrypoint) > 128 || len(call.Calldata) > 1048576 || len(call.Capabilities) > 65535 || len(call.AccessDeclaration) > 1048576 || call.ResponseCapacity > 1048576 {
		return nil, newSDKError(ErrorInvalidArgument, RetryNever)
	}
	for _, b := range []byte(call.Entrypoint) {
		if !(b >= 'a' && b <= 'z' || b >= 'A' && b <= 'Z' || b >= '0' && b <= '9' || b == '_' || b == '.') {
			return nil, newSDKError(ErrorInvalidArgument, RetryNever)
		}
	}
	out := make([]byte, 106, 106+len(call.Entrypoint)+len(call.Calldata)+len(call.Capabilities)+len(call.AccessDeclaration))
	copy(out, call.ProgramID[:])
	binary.BigEndian.PutUint16(out[32:], call.GuestABI)
	binary.BigEndian.PutUint16(out[34:], uint16(len(call.Entrypoint)))
	binary.BigEndian.PutUint32(out[36:], uint32(len(call.Calldata)))
	binary.BigEndian.PutUint16(out[40:], uint16(len(call.Capabilities)))
	binary.BigEndian.PutUint32(out[42:], uint32(len(call.AccessDeclaration)))
	binary.BigEndian.PutUint32(out[46:], call.ResponseCapacity)
	for i, v := range call.Resources {
		binary.BigEndian.PutUint64(out[50+i*8:], v)
	}
	for _, body := range [][]byte{[]byte(call.Entrypoint), call.Calldata, call.Capabilities, call.AccessDeclaration} {
		out = append(out, body...)
	}
	return out, nil
}

func DecodeNativeProgramCall(payload []byte) (NativeProgramCall, error) {
	var call NativeProgramCall
	invalid := newSDKError(ErrorInvalidArgument, RetryNever)
	if len(payload) < 106 {
		return call, invalid
	}
	lengths := []uint64{uint64(binary.BigEndian.Uint16(payload[34:])), uint64(binary.BigEndian.Uint32(payload[36:])), uint64(binary.BigEndian.Uint16(payload[40:])), uint64(binary.BigEndian.Uint32(payload[42:]))}
	total := uint64(106)
	for _, n := range lengths {
		total += n
	}
	if total != uint64(len(payload)) {
		return call, invalid
	}
	copy(call.ProgramID[:], payload[:32])
	call.GuestABI = binary.BigEndian.Uint16(payload[32:])
	call.ResponseCapacity = binary.BigEndian.Uint32(payload[46:])
	for i := range call.Resources {
		call.Resources[i] = binary.BigEndian.Uint64(payload[50+i*8:])
	}
	bodies := make([][]byte, 4)
	offset := 106
	for i, n := range lengths {
		bodies[i] = append([]byte(nil), payload[offset:offset+int(n)]...)
		offset += int(n)
	}
	call.Entrypoint = string(bodies[0])
	call.Calldata = bodies[1]
	call.Capabilities = bodies[2]
	call.AccessDeclaration = bodies[3]
	if _, err := EncodeNativeProgramCall(call); err != nil {
		return NativeProgramCall{}, err
	}
	return call, nil
}

type NativeProgramRequest struct {
	Call           NativeProgramCall
	FeeLimit       Uint128
	SignedActivity []byte
}

func (request NativeProgramRequest) ProgramCall() (ProgramCall, error) {
	if _, err := EncodeNativeProgramCall(request.Call); err != nil {
		return ProgramCall{}, err
	}
	call := ProgramCall{NativeCall: &request.Call, ProgramID: request.Call.ProgramID, Calldata: request.Call.Calldata, Budget: ProgramBudget{Fuel: request.Call.Resources[0], FeeLimit: request.FeeLimit}, SignedActivity: request.SignedActivity}
	if err := validateProgramCall(call); err != nil {
		return ProgramCall{}, err
	}
	return call, nil
}

File: platform/sdk/dotnet/NativeProgramCall.cs
using System.Buffers.Binary;
using System.Text;
using System.Text.RegularExpressions;

namespace LayerX.Sdk;

public sealed record NativeProgramCall(byte[] ProgramId, ushort GuestAbi, string Entrypoint, byte[] Calldata,
    byte[] Capabilities, byte[] AccessDeclaration, uint ResponseCapacity, ulong[] Resources)
{
    public byte[] Encode()
    {
        if (ProgramId.Length != 32 || ProgramId.All(value => value == 0) || GuestAbi is not (1 or 2)
            || !Regex.IsMatch(Entrypoint, "\\A[A-Za-z0-9_.]{1,128}\\z") || Calldata.Length > 1048576
            || Capabilities.Length > 65535 || AccessDeclaration.Length > 1048576 || ResponseCapacity > 1048576 || Resources.Length != 7)
            throw new ArgumentException("invalid native program call");
        var entry = Encoding.ASCII.GetBytes(Entrypoint);
        var output = new byte[106 + entry.Length + Calldata.Length + Capabilities.Length + AccessDeclaration.Length];
        ProgramId.CopyTo(output, 0); BinaryPrimitives.WriteUInt16BigEndian(output.AsSpan(32), GuestAbi);
        BinaryPrimitives.WriteUInt16BigEndian(output.AsSpan(34), (ushort)entry.Length);
        BinaryPrimitives.WriteUInt32BigEndian(output.AsSpan(36), (uint)Calldata.Length);
        BinaryPrimitives.WriteUInt16BigEndian(output.AsSpan(40), (ushort)Capabilities.Length);
        BinaryPrimitives.WriteUInt32BigEndian(output.AsSpan(42), (uint)AccessDeclaration.Length);
        BinaryPrimitives.WriteUInt32BigEndian(output.AsSpan(46), ResponseCapacity);
        for (var i = 0; i < 7; i++) BinaryPrimitives.WriteUInt64BigEndian(output.AsSpan(50 + i * 8), Resources[i]);
        var offset = 106;
        foreach (var body in new[] { entry, Calldata, Capabilities, AccessDeclaration }) { body.CopyTo(output, offset); offset += body.Length; }
        return output;
    }

    public static NativeProgramCall Decode(byte[] payload)
    {
        if (payload.Length < 106) throw new ArgumentException("invalid native program call");
        var sizes = new ulong[] { BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(34)), BinaryPrimitives.ReadUInt32BigEndian(payload.AsSpan(36)), BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(40)), BinaryPrimitives.ReadUInt32BigEndian(payload.AsSpan(42)) };
        if (106UL + sizes[0] + sizes[1] + sizes[2] + sizes[3] != (ulong)payload.Length) throw new ArgumentException("invalid native program call");
        var resources = new ulong[7]; for (var i = 0; i < 7; i++) resources[i] = BinaryPrimitives.ReadUInt64BigEndian(payload.AsSpan(50 + i * 8));
        var bodies = new byte[4][]; var offset = 106;
        for (var i = 0; i < 4; i++) { bodies[i] = payload.AsSpan(offset, (int)sizes[i]).ToArray(); offset += (int)sizes[i]; }
        var call = new NativeProgramCall(payload[..32], BinaryPrimitives.ReadUInt16BigEndian(payload.AsSpan(32)), Encoding.ASCII.GetString(bodies[0]), bodies[1], bodies[2], bodies[3], BinaryPrimitives.ReadUInt32BigEndian(payload.AsSpan(46)), resources);
        if (!call.Encode().SequenceEqual(payload)) throw new ArgumentException("invalid native program call");
        return call;
    }
}
